lazy_load_symbols/load_sym_lazily.py

```diff
@@ -0,0 +1,171 @@
+                    self.sodict[name]['end'] = ea
+                    self.sodict[name]['loaded'] = False
+                else:
+                    # this shouldn't happen but we can just verify it is indeed the same library
+                    assert self.sodict[name]['start'] == sa, 'found same' \
+                        ' library ({}) at 2 different locations: {} != {}' \
+                        .format(name, sa, self.sodict[name]['start'])
+                    assert self.sodict[name]['end'] == ea, 'found same library' \
+                        ' ({}) ending at 2 different locations: {} != {}' \
+                        .format(name, ea, self.sodict[name]['end'])
+        return self.sodict
+
+    def load_all_symbols(self):
+        for so in self.sodict:
+            fn = self.sodict[so]['fullname']
+            nm = Popen(['nm', '-D', fn], stdout=PIPE)
+            (symlist, err) = nm.communicate()
+            assert not err, 'nm failed somehow {}'.format(err)
+            assert symlist, 'no symbols in file {}'.format(fn)
+            ec = nm.wait()
+            assert not ec, 'nm exited with {}'.format(ec)
+            self.sodict[so]['symbols'] = symlist
+
+    def invoke(self, rec, from_tty):
+        if self.sodict:
+            pass
+        else:
+            gdb.execute('set auto-solib-add off', to_string=True)
+            gdb.execute('uload {}'.format(rec))
+
+
+class PopulateBT(gdb.Command):
+    '''
+    A command to read the symbols for an empty backtrace
+    '''
+
+    def __init__(self):
+        super(PopulateBT, self).__init__('popbt', gdb.COMPLETE_EXPRESSION)
+        self.sodict = {}
+
+    def get_frame_name(self, pc):
+        for name in self.sodict:
+            sa = int(self.sodict[name]['start'], 16)
+            ea = int(self.sodict[name]['end'], 16)
+            if pc >= sa and pc < ea:
+                gdb.execute('sharedlibrary {}'.format(name))
+
+    def invoke(self, arg, from_tty):
+        ll = LoadLazySymbols()
+        self.sodict = ll.load_so_list()
+        curr_frame = gdb.newest_frame()
+        while curr_frame:
+            try:
+                # iteratively go through all frames and for each one check if the name is loaded.
+                # if not load the name
+                if not curr_frame.name():
+                    self.get_frame_name(curr_frame.pc())
+                curr_frame = curr_frame.older()
+            except gdb.error:
+                # restart as loading frames might change things ;)
+                curr_frame = gdb.newest_frame()
+        self.sodict = {}
+
+
+class LoadLibFromSym(gdb.Command):
+    '''
+    Command to load all libs that might define symbol
+    '''
+
+    def __init__(self):
+        super(LoadLibFromSym, self).__init__('loadsymlib', gdb.COMPLETE_EXPRESSION)
+        self.sodict = {}
+
+    def read_lib_symbols(self, so):
+        fn = self.sodict[so]['fullname']
+        nm = Popen(['nm', '-D', fn], stdout=PIPE)
```

Comment:
`subprocess.check_output`.
Or maybe this code can be shared with the other code calling `nm`?

Reply:
yup, good idea, going to do that (check_output AND creating a function out of this code)